pkg/gitfs/fs.go

```diff
@@ -0,0 +1,172 @@
+package gitfs
+
+import (
+	"fmt"
+	"io"
+	"io/fs"
+	"path"
+	"strings"
+	"time"
+
+	goGit "github.com/go-git/go-git/v5"
+	goGitPlumbing "github.com/go-git/go-git/v5/plumbing"
+	goGitPlumbingFileMode "github.com/go-git/go-git/v5/plumbing/filemode"
+	goGitPlumbingObject "github.com/go-git/go-git/v5/plumbing/object"
+)
+
+// https://github.com/go-git/go-git/issues/296
+
+// TODO something more clever for directories
+
+func CommitHash(repo *goGit.Repository, commit string) (fs.FS, error) {
+	gitCommit, err := repo.CommitObject(goGitPlumbing.NewHash(commit))
+	if err != nil {
+		return nil, err
+	}
+	return gitFS{
+		commit: gitCommit,
+	}, nil
+}
+
+// https://pkg.go.dev/io/fs#FS
+type gitFS struct {
+	commit *goGitPlumbingObject.Commit
+}
+
+// apparently symlinks in "io/fs" are still a big TODO (https://github.com/golang/go/issues/49580, https://github.com/golang/go/issues/45470, etc related issues); all the existing interfaces assume symlinks don't exist
+//
+// if the File object passed to this function represents a symlink, this returns the (resolved) path that should be looked up instead; only relative symlinks are supported (and attempts to escape the repository with too many "../" *should* result in an error -- this is a convenience/sanity check, not a security boundary; subset of https://pkg.go.dev/io/fs#ValidPath)
+//
+// otherwise, it will return the empty string and nil
+func resolveSymlink(f *goGitPlumbingObject.File) (target string, err error) {
+	if f.Mode != goGitPlumbingFileMode.Symlink {
+		return "", nil
+	}
+
+	target, err = f.Contents()
+	if err != nil {
+		return "", err
+	}
+
+	if target == "" {
+		return "", fmt.Errorf("unexpected: empty symlink %q", f.Name)
+	}
+
+	if path.IsAbs(target) {
+		return "", fmt.Errorf("unsupported: %q is an absolute symlink (%q)", f.Name, target)
+	}
+
+	target = path.Join(path.Dir(f.Name), target)
+
+	if strings.HasPrefix(target, "../") {
+		return "", fmt.Errorf("unsupported: %q is a relative symlink outside the tree (%q)", f.Name, target)
+	}
```

Comment:
This probably should be above the `path.Join`? And maybe a check on `path.Clean`?

Reply:
It needs to be below the `path.Join` because a relative symlink to a directory outside of its directory but still within the Git repository is valid and safe (it's only leaving the Git repository that isn't safe and would actually error later so this is still a best-effort check).  I'll add some comments to make that more clear.

That being said, `path.Clean` is a good call-out - that would make this check's implementation safer. :sweat_smile: